safenode/src/storage/mod.rs

```diff
@@ -6,53 +6,68 @@
 // KIND, either express or implied. Please review the Licences for the specific language governing
 // permissions and limitations relating to use of the SAFE Network Software.
 
-/// Chunks
-pub mod chunks;
+use crate::protocol::{
+    messages::{Query, QueryResponse, ReplicatedData},
+    types::{address::ChunkAddress, chunk::Chunk, register::User},
+};
+
+mod chunks;
 mod errors;
+mod register_store;
+mod registers;
 
-use self::chunks::{Chunk, ChunkAddress};
 use chunks::ChunkStorage;
-use errors::Result;
-use std::path::{Path, PathBuf};
-use xor_name::XorName;
-
-const BIT_TREE_DEPTH: usize = 20;
+use errors::{Error, Result};
+use registers::RegisterStorage;
+use tracing::debug;
 
 /// Operations on data stored to disk.
 /// As data the storage struct may be cloned throughoout the node
 /// Operations here must be persisted to disk.
-#[derive(Debug, Clone)]
-// exposed as pub due to benches
+// Exposed as pub due to benches.
+#[derive(Clone, Default)]
 pub struct DataStorage {
     chunks: ChunkStorage,
+    registers: RegisterStorage,
 }
 
 impl DataStorage {
     /// Set up a new `DataStorage` instance
-    pub fn new(path: &Path) -> Self {
+    pub fn new() -> Self {
         Self {
-            chunks: ChunkStorage::new(path),
+            chunks: ChunkStorage::default(),
+            registers: RegisterStorage::new(),
         }
     }
 
-    /// Store data in the local store
-    pub async fn store(&self, chunk: &Chunk) -> Result<()> {
+    /// Store Chunk in the local store
+    pub async fn store_chunk(&self, chunk: &Chunk) -> Result<()> {
         self.chunks.store(chunk).await
     }
 
     /// Query the local store and return the Chunk
-    pub async fn query(&self, addr: &ChunkAddress) -> Result<Chunk> {
+    pub async fn query_chunk(&self, addr: &ChunkAddress) -> Result<Chunk> {
         self.chunks.get(addr).await
     }
-}
 
-// Helper that returns the prefix tree path of depth BIT_TREE_DEPTH for a given xorname
-// Example:
-// - with a xorname with starting bits `010001110110....`
-// - and a BIT_TREE_DEPTH of `6`
-// returns the path `ROOT_PATH/0/1/0/0/0/1`
-fn prefix_tree_path(root: &Path, xorname: XorName) -> PathBuf {
-    let bin = format!("{xorname:b}");
-    let prefix_dir_path: PathBuf = bin.chars().take(BIT_TREE_DEPTH).map(String::from).collect();
-    root.join(prefix_dir_path)
+    /// Store data in the local store
+    pub async fn store(&self, data: &ReplicatedData) -> Result<()> {
+        debug!("Replicating {data:?}");
+        match data {
+            ReplicatedData::Chunk(chunk) => self.chunks.store(chunk).await,
+            ReplicatedData::RegisterLog(data) => self.registers.update(data).await,
+            ReplicatedData::RegisterWrite(cmd) => self.registers.write(cmd).await,
+        }
+    }
+
+    /// Query the local store and return QueryResponse
+    pub async fn query(&self, query: &Query, requester: User) -> QueryResponse {
```

Comment:
Since we take a query here, maybe we should take a cmd above, for stringency?

That would also imply that we implement and keep spentbook/dbc spend storage in DataStorage struct.

But I guess that is implied if we take Query here, and will keep dbc requests as part of the Data reqquests.